lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -4273,6 +4276,46 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
     return BM->addExtInst(FTy, BM->getExtInstSetId(SPIRVEIS_OpenCL), ExtOp, Ops,
                           BB);
   }
+  case Intrinsic::modf: {
+    // llvm.modf has a single arg --the number to be decomposed-- and returns a
+    // struct { double, double }, while OpenCLLIB::modf has two args --the
+    // number to be decomposed and a pointer--, returns the fractional part and
+    // the integral part is stored in the pointer argument. Therefore, we can't
+    // use directly the OpenCLLIB::modf intrinsic. However, we can do some
+    // scaffolding to make it work. The idea is to create an alloca instruction
+    // to get a ptr, pass this ptr to OpenCLLIB::modf, and then load the value
+    // from this ptr to place it in the struct.
+
+    // llvm.modf returns the fractional part as the first element of the result,
+    // and the integral part as the second element of the result. Therefore, the
+    // first element is the return value of OpenCLLIB::modf, and the second
+    // element is the value loaded from the ptr of the alloca we created.
+
+    // Create the alloca instruction.
+    SPIRVType *IntegralTy = transType(II->getType()->getStructElementType(1));
+    // IntegralTy is the type of the result. We want to create a pointer to this
+    // that we can pass to OpenCLLIB::modf to store the integral part.
+    SPIRVTypePointer *IntegralPtrTy =
+        BM->addPointerType(StorageClassFunction, IntegralTy);
+    SPIRVValue *Ptr = BM->addVariable(IntegralPtrTy, nullptr, false,
+                                      spv::internal::LinkageTypeInternal,
+                                      nullptr, "", StorageClassFunction, BB);
```

Comment:
Ideally it should be not current BB, but entry BB of the functions to satisfy "All [OpVariable](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpVariable) instructions in a function must have a [Storage Class](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Storage_Class) of Function.".

SPIRVFunction class in the translator has BB vector, so it might be possible to query first basic block and then call `getVariableInsertionPoint` to get a place where to insert `OpVariable` (or alternatively just create alloca in the beginning of the function using LLVM API and call `transValue` here).